sewer/dns_providers/acmedns.py

```diff
@@ -0,0 +1,69 @@
+import urllib.parse
+
+from dns.resolver import Resolver
+import requests
+
+from . import common
+
+
+class AcmeDnsDns(common.BaseDns):
+    """
+    """
+    dns_provider_name = 'acmedns'
+
+    def __init__(
+            self,
+            ACME_DNS_API_USER,
+            ACME_DNS_API_KEY,
+            ACME_DNS_API_BASE_URL):
+        self.ACME_DNS_API_USER = ACME_DNS_API_USER
+        self.ACME_DNS_API_KEY = ACME_DNS_API_KEY
+        self.HTTP_TIMEOUT = 65  # seconds
+
+        if ACME_DNS_API_BASE_URL[-1] != '/':
+            self.ACME_DNS_API_BASE_URL = ACME_DNS_API_BASE_URL + '/'
+        else:
+            self.ACME_DNS_API_BASE_URL = ACME_DNS_API_BASE_URL
+        super(AcmeDnsDns, self).__init__()
+
+    def create_dns_record(self, domain_name, domain_dns_value):
+        self.logger.info('create_dns_record')
+        # if we have been given a wildcard name, strip wildcard
+        domain_name = domain_name.lstrip('*.')
+
+        resolver = Resolver(configure=False)
+        resolver.nameservers = ['8.8.8.8']
+        answer = resolver.query('_acme-challenge.{0}.'.format(domain_name), 'TXT')
+        subdomain, _ = str(answer.canonical_name).split('.', 1)
+
```

Comment:
@dnet Hi, do we need `dnspython` in order to get the value of `subdomain`? 

I do not really know how acme-dns works, but going through its documentation;
wouldn't it be possible to get the value of `subdomain` just by making a `POST` request to the registration endpoint? ie;
```python
url = urllib.parse.urljoin(self.ACME_DNS_API_BASE_URL, 'register')
resp = requests.get(url, headers=headers, timeout=self.HTTP_TIMEOUT)
subdomain = resp.json()['subdomain']
```
and then use the subdomain in the subsequent request to the `update` endpoint

Reply:
I don't like the idea of the dependency either, however the main use-case of _acme-dns_ is pointing a `CNAME` record to the _acme-dns_ instance from the main domain. So for example, in case of `hsbp.org`, I've set it up like this: (`auth.vsza.hu` is my private _acme-dns_ instance)

```
$ host -a _acme-challenge.hsbp.org.
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 41523
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;_acme-challenge.hsbp.org.      IN      ANY

;; ANSWER SECTION:
_acme-challenge.hsbp.org. 3599  IN      CNAME   78d4dd08-c13d-4700-acdc-2bd2a7b3f2ef.auth.vsza.hu.
```

You're right, I got `78d4dd08-c13d-4700-acdc-2bd2a7b3f2ef` from the response to the `register` call, however, I did that only once, as the whole point that _acme-dns_ solves is that I don't want to update the `_acme-challenge` entry, I just want to set it up once to point to _acme-dns. And it seemed to me that the Python standard library doesn't have the bits to perform the necessary DNS queries, since `gethostbyname{,_ex}` relies on `A` and `AAAA` records only.

If you really dislike the dependency, I see several ways, some of them could even be combined:

 - Put it as an _extra_ dependency, since it's only used by the _acme-dns_ module (and let's face it, the zeitgeist is people using cloud magic, self-hosters like me are a minority, even if a vocal one).
 - Fall back to calling `host(1)`, `nslookup(1)` or `dig(1)` using `subprocess` (ugly but it can work).
 - Let the user specify the `subdomain` (like after 6ead878 but before 85aabfe), although this might cause some problems if there are multiple domains specified using `domain_alt_names` and not all might point to the same _acme-dns_ instance.

Reply:
I think we can keep the dependency then since the alternatives do not appear to be much better. 
At a later date we may refactor and put `dns-python` as an extra dependency.